### PR TITLE
Add refresh token

### PR DESCRIPTION
**Pull request by michaelKaefer**
We need a refresh token in our app.

### CODE REVIEW

File: src/OidcClient.php [hunk 1]
```diff
@@ -99,6 +99,30 @@ public function authenticate(Request $request): OidcTokens
     }
   }
 
+  public function refreshTokens(string $refreshToken): OidcTokens
```
Comment (bobvandevijver) [inline src/OidcClient.php:102]: This method looks to be mainly a duplication of the already existing authenticate method. I would love to see a refactor so we have as less duplication as possible.

File: src/OidcClient.php [hunk 2]
```diff
@@ -266,15 +290,12 @@ private function getConfigurationValue(string $key): mixed
    *
    * @throws OidcException
    */
-  private function requestTokens(string $code): OidcTokens
+  private function requestTokens(array $params): OidcTokens
```
Comment (bobvandevijver) [inline src/OidcClient.php:293]: I'm in doubt about this change. I think I prefer the method to be as strictly defined as possible, so something like:

```php
private function requestTokens(string $grantType, string $code = null, string $refreshToken = null, string $redirectUrl = null): OidcTokens
```

That way, we can utilise named parameters in the call, and only add the param when it has been set.

What do you think?

File: src/OidcJwtHelper.php [hunk 3]
```diff
@@ -94,17 +94,19 @@ public function verifyJwtClaims(string $issuer, ?object $claims, ?OidcTokens $to
...
+    if (true === $verifyNonce) {
+      $nonce = $this->sessionStorage->getNonce();
+      $this->sessionStorage->clearNonce();
+    }
 
     /* @noinspection PhpUndefinedVariableInspection */
     return ($claims->iss == $issuer)
         && (($claims->aud == $this->clientId) || (in_array($this->clientId, $claims->aud)))
-        && ($claims->nonce == $nonce)
+        && (false === $verifyNonce || $claims->nonce == $nonce)
```
Comment (bobvandevijver) [inline src/OidcJwtHelper.php:105]:
```suggestion
        && (!$verifyNonce || $claims->nonce == $nonce)
```

File: src/OidcJwtHelper.php [hunk 4]
```diff
@@ -94,17 +94,19 @@ public function verifyJwtClaims(string $issuer, ?object $claims, ?OidcTokens $to
...
         && (!isset($claims->exp) || $claims->exp >= time())
         && (!isset($claims->nbf) || $claims->nbf <= time())
         && (!isset($claims->at_hash) || $claims->at_hash == $expectedAtHash)
-    ;
+        ;
```
Comment (bobvandevijver) [inline src/OidcJwtHelper.php:109]:
```suggestion
    ;
```

File: src/OidcJwtHelper.php (hunk 3 above)
Comment (bobvandevijver) [reply src/OidcJwtHelper.php:105]: We utilize a private pipeline in our main repository, which complains about the `$nonce` here as well:

```
src/OidcJwtHelper.php:105 PhanPossiblyUndeclaredVariable Variable $nonce is possibly undeclared
```

See my other suggestion on line 97 how I think we can best fix this.

File: src/OidcJwtHelper.php [hunk 5]
```diff
@@ -94,17 +94,19 @@ public function verifyJwtClaims(string $issuer, ?object $claims, ?OidcTokens $to
     }
 
     // Get and remove nonce from session
-    $nonce = $this->sessionStorage->getNonce();
-    $this->sessionStorage->clearNonce();
+    if (true === $verifyNonce) {
+      $nonce = $this->sessionStorage->getNonce();
+      $this->sessionStorage->clearNonce();
+    }
```
Comment (bobvandevijver) [inline src/OidcJwtHelper.php:100]:
```suggestion
    $nonce = $verifyNonce ? $this->sessionStorage->getNonce() : null;
    if (null !== $nonce) {
      $this->sessionStorage->clearNonce();
    }
```

**Review by bobvandevijver — CHANGES_REQUESTED** [review]
Thank you for your PR, it looks great! I have not method of testing this with my current OIDC providers (they do not support refresh tokens), so I will have to trust you on whether this works correctly.

I just have a couple of small suggestions, hopefully you can do something with them.

File: src/OidcClient.php (hunk 1 above)
Comment (michaelKaefer) [reply src/OidcClient.php:102]: @bobvandevijver For me it is ok, the duplication makes sense for me. Moving it into a seperate method would make it harder to read and provides little improvement IMHO.

Comment (bobvandevijver) [reply src/OidcClient.php:102]: As it is quite a lot of shared implementation, I have now refactored it into a `verifyTokens` method.

**Comment by bobvandevijver** [comment]
Merged with c99a9cf3e9a5553218fcfb2cf0af0e6a6ba5eb35 & 3038a2df0b119b3d7c055bfbedfcdfba662c721c.

**Pull request closed without merging**